### DISCUSSION TITLE
Websocket endpoint returns 403 Forbidden after 0.0.11 upgrade

### DISCUSSION THREAD

**Discussion by creideiki** [owncast/owncast#1758] [Q&A] [answered]
After upgrading to 0.0.11, chat does not work in my Owncast server, but it did work in 0.0.10.

It sits behind an Apache TLS-terminating reverse proxy, configured according to https://owncast.online/docs/sslproxies/apache/

Sniffing the (unencrypted) traffic between Apache and Owncast, it seems Owncast doesn't like upgrading websockets and returns 403 Forbidden:

Request (from Apache to Owncast):
```
GET /ws?accessToken=zNsDWaAe1v3mVt-beLymhiE1m_0YmEfeG5Iq_fyh5Jo%3D HTTP/1.1
Host: 192.168.0.40:8080
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:97.0) Gecko/20100101 Firefox/97.0
Accept: */*
Accept-Language: en-GB,en-US;q=0.8,en;q=0.6,sv-SE;q=0.4,sv;q=0.2
Accept-Encoding: gzip, deflate, br
Sec-WebSocket-Version: 13
Origin: https://owncast.ferretporn.se
Sec-WebSocket-Extensions: permessage-deflate
Sec-WebSocket-Key: jaYc9tW2veZKji+iaLK+6A==
DNT: 1
Sec-Fetch-Dest: websocket
Sec-Fetch-Mode: websocket
Sec-Fetch-Site: same-origin
Pragma: no-cache
Cache-Control: no-cache
X-Forwarded-Proto: https
X-Forwarded-Port: 443
X-Forwarded-For: 192.168.3.238
X-Forwarded-Host: owncast.ferretporn.se
X-Forwarded-Server: owncast.ferretporn.se
Connection: Upgrade
Upgrade: websocket
```
Response (from Owncast to Apache):
```
HTTP/1.1 403 Forbidden
Content-Type: text/plain; charset=utf-8
Sec-Websocket-Version: 13
X-Content-Type-Options: nosniff
Date: Sun, 06 Mar 2022 12:02:41 GMT
Content-Length: 10

Forbidden
```
Are there any changes necessary to make this setup work in 0.0.11?

**Comment by gabek** [comment]
There are no changes required, and it seems to be working for others. I only know of one person using Apache as a proxy, but it's been working for him, as well. So unfortunately I'm not sure why you're seeing what you're seeing.  If I had to guess your proxy is removing the access token, so Owncast is rejecting the connection.

**Reply by creideiki** [reply]
> If I had to guess your proxy is removing the access token, so Owncast is rejecting the connection.

The traffic dump is from the network behind the proxy, so it shows exactly what Owncast saw.

**Comment by creideiki** [comment]
No change after upgrading to 0.0.12. The reverse proxy configuration is still what [the documentation](https://owncast.online/docs/sslproxies/apache/) recommends:
```
        RequestHeader set X-Forwarded-Proto "https"
        RequestHeader set X-Forwarded-Port "443"

        RewriteEngine On
        RewriteCond %{HTTP:UPGRADE} WebSocket  [NC,OR]
        RewriteCond %{HTTP:CONNECTION} Upgrade  [NC]
        RewriteRule .* ws://192.168.0.40:8080%{REQUEST_URI}  [P,QSA,L]
```
The reverse proxy still forwards the access token:
```
GET /ws?accessToken=zNsDWaAe1v3mVt-beLymhiE1m_0YmEfeG5Iq_fyh5Jo%3D HTTP/1.1
Host: 192.168.0.40:8080
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:102.0) Gecko/20100101 Firefox/102.0
Accept: */*
Accept-Language: en-GB,en-US;q=0.8,en;q=0.6,sv-SE;q=0.4,sv;q=0.2
Accept-Encoding: gzip, deflate, br
Sec-WebSocket-Version: 13
Origin: https://owncast.ferretporn.se
Sec-WebSocket-Extensions: permessage-deflate
Sec-WebSocket-Key: 4CSP3q7xETW6v5Vlng4w1Q==
DNT: 1
Sec-Fetch-Dest: websocket
Sec-Fetch-Mode: websocket
Sec-Fetch-Site: same-origin
Pragma: no-cache
Cache-Control: no-cache
X-Forwarded-Proto: https
X-Forwarded-Port: 443
X-Forwarded-For: 2001:470:dd6d:3:b3c5:a664:48ff:fcbf
X-Forwarded-Host: owncast.ferretporn.se
X-Forwarded-Server: owncast.ferretporn.se
Connection: Upgrade
Upgrade: websocket
```
and Owncast still replies immediately with 403 Forbidden:
```
HTTP/1.1 403 Forbidden
Content-Type: text/plain; charset=utf-8
Sec-Websocket-Version: 13
X-Content-Type-Options: nosniff
Date: Sat, 16 Jul 2022 18:03:10 GMT
Content-Length: 10

Forbidden
```

**Reply by gabek** [reply]
Our proxy documentation is generally community supplied, so while it's there to hopefully get you running, I don't use Apache as a proxy myself and highly suggest reading the official Apache documentation if you're running into issues. If Owncast is working without a proxy then you're in good shape and just need to work our your Apache configuration, something, I'm afraid, I can't supply for you if the example in the documentation isn't already working.

**Comment by creideiki** [comment, marked as answer]
I was sufficiently annoyed by this that I cargo-culted up some `printf` debugging, and I think I know what's going on now.

The connection fails when `HandleClientConnection` tries to upgrade to a websocket by calling `upgrader.Upgrade`, here: https://github.com/owncast/owncast/blob/fb3973b264d82aad7e0ae3086255d72506a38c41/core/chat/server.go#L187
The error message is "Upgrade error: websocket: request origin not allowed by Upgrader.CheckOrigin".

Owncast does not set a `CheckOrigin` function when it creates the Upgrader: https://github.com/owncast/owncast/blob/fb3973b264d82aad7e0ae3086255d72506a38c41/core/chat/chatclient.go#L60
So then, the websocket library uses a default function: https://github.com/gorilla/websocket/blob/af47554f343b4675b30172ac301638d350db34a5/server.go#L149
That checks if the `Host` and `Origin` HTTP headers match: https://github.com/gorilla/websocket/blob/af47554f343b4675b30172ac301638d350db34a5/server.go#L88

While this may be an appropriate (but not especially meaningful) test for a server directly connected to the Internet, after going through a reverse proxy, the `Host` and `Origin` headers shouldn't ever match, since `Origin` should be the outside address of the reverse proxy, and `Host` should be the internal address of Owncast. This is the case with my dumped traffic above:
```
Host: 192.168.0.40:8080
Origin: https://owncast.ferretporn.se
```
Note that this is not in general fixable by configuration in the reverse proxy, since it's the host part of the `Origin` header that must match the entire `Host` header, and since the `Host` header includes a port number, the host part of the `Origin` header can never match it.

I would think that a better test would be to check whether the `Origin` header has the same host as the Server URL that Owncast has been configured with, since both of those should be the outside address of the reverse proxy.

Unfortunately, here's where my cargo-culted Go code let me down, and I couldn't get that to work. Instead, I wrote a custom `CheckOrigin` function that just returns `true`:
```patch
diff --git a/core/chat/chatclient.go b/core/chat/chatclient.go
index fdcd77e3..4408bd48 100644
--- a/core/chat/chatclient.go
+++ b/core/chat/chatclient.go
@@ -3,6 +3,7 @@ package chat
 import (
        "bytes"
        "encoding/json"
+       "net/http"
        "sync"
        "time"
 
@@ -60,6 +61,9 @@ const (
 var upgrader = websocket.Upgrader{
        ReadBufferSize:  1024,
        WriteBufferSize: 1024,
+       CheckOrigin:     func(r *http.Request) bool {
+               return true
+       },
 }
 
 var (
```
With that change in place, chat works for me.

I'm not sure if that may even be appropriate for production, since the access token check should be enough to prevent request forgery, which seems like a stronger check than just looking at if two client-provided header values match.

**Reply by gabek** [reply]
Makes sense! I think most people's configurations rewrite the host header, and that's why I've never heard of anybody else running into this. I think Caddy sets it automatically, and in the Nginx documentation it's a part of the config. So you could probably set it with Apache as well if you wanted to. But I don't see any reason we don't do like you suggested and just force the origin check to pass because of exactly what you said, the access token is the primary check anyway. Plus if somebody wanted to build some kind of custom websocket powered thing elsewhere there's no reason this check should stop them.

Thanks for looking into it!

**Reply by gabek** [reply]
Filed #2003.
